Add an AccessPath abstraction and formalize memory access

atrick opened the pull request:
Things that have come up recently but are somewhat blocked on this:

    - Moving AccessMarkerElimination down in the pipeline
    - SemanticARCOpts correctness and improvements
    - AliasAnalysis improvements
    - LICM performance regressions
    - RLE/DSE improvements

    Begin to formalize the model for valid memory access in SIL. Ignoring
    ownership, every access is a def-use chain in three parts:

    object root -> formal access base -> memory operation address

    AccessPath abstracts over this path and standardizes the identity of a
    memory access throughout the optimizer. This abstraction is the basis
    for a new AccessPathVerification.

    With that verification, we now have all the properties we need for the
    type of analysis requires for exclusivity enforcement, but now
    generalized for any memory analysis. This is suitable for an extremely
    lightweight analysis with no side data structures. We currently have a
    massive amount of ad-hoc memory analysis throughout SIL, which is
    incredibly unmaintainable, bug-prone, and not performance-robust. We
    can begin taking advantage of this verifably complete model to solve
    that problem.

    The properties this gives us are:

    Access analysis must be complete over memory operations: every memory
    operation needs a recognizable valid access. An access can be
    unidentified only to the extent that it is rooted in some non-address
    type and we can prove that it is at least *not* part of an access to a
    nominal class or global property. Pointer provenance is also required
    for future IRGen-level bitfield optimizations.

    Access analysis must be complete over address users: for an identified
    object root all memory accesses including subobjects must be
    discoverable.

    Access analysis must be symmetric: use-def and def-use analysis must
    be consistent.

    AccessPath is merely a wrapper around the existing accessed-storage
    utilities and IndexTrieNode. Existing passes already very succesfully
    use this approach, but in an ad-hoc way. With a general utility we
    can:

    - update passes to use this approach to identify memory access,
      reducing the space and time complexity of those algorithms.

    - implement an inexpensive on-the-fly, debug mode address lifetime analysis

    - implement a lightweight debug mode alias analysis

    - ultimately improve the power, efficiency, and maintainability of
      full alias analysis

    - make our type-based alias analysis sensistive to the access path

atrick committed:
Change Projection to support signed indices.

Change ProjectionIndex for ref_tail_addr to std::numeric_limits<int>::max();
This is necessary to disambiguate the tail elements from
ref_element_addr field zero.

atrick committed:
Add indexTrieRoot to the SILModule to share across Analyses.

...and avoid reallocation.

This is immediately necessary for LICM, in addition to its current
uses. I suspect this could be used by many passes that work with
addresses. RLE/DSE should absolutely migrate to it.

atrick committed:
Fix a pointer addition bug in SwiftReflectionTest.

atrick committed:
Change IndexTrie to allow signed indices

atrick committed:
Fix ValueTracking isUniquelyIdentified to use AccessedStorage.

To clarify and unify logic, improve precision, and behave consistently
with other code that does the same thing.

atrick committed:
Add AccessPathVerification pass and run it in the pipeline.

atrick committed:
Add AccessPath support to the AccessedStorageDumper pass.

Add a flag: enable-accessed-storage-dump-uses

atrick committed:
accessed_storage.sil tests

atrick committed:
accesspath_uses.sil tests

atrick committed:
Add an AccessPath abstraction and formalize memory access

Things that have come up recently but are somewhat blocked on this:

- Moving AccessMarkerElimination down in the pipeline
- SemanticARCOpts correctness and improvements
- AliasAnalysis improvements
- LICM performance regressions
- RLE/DSE improvements

Begin to formalize the model for valid memory access in SIL. Ignoring
ownership, every access is a def-use chain in three parts:

object root -> formal access base -> memory operation address

AccessPath abstracts over this path and standardizes the identity of a
memory access throughout the optimizer. This abstraction is the basis
for a new AccessPathVerification.

With that verification, we now have all the properties we need for the
type of analysis requires for exclusivity enforcement, but now
generalized for any memory analysis. This is suitable for an extremely
lightweight analysis with no side data structures. We currently have a
massive amount of ad-hoc memory analysis throughout SIL, which is
incredibly unmaintainable, bug-prone, and not performance-robust. We
can begin taking advantage of this verifably complete model to solve
that problem.

The properties this gives us are:

Access analysis must be complete over memory operations: every memory
operation needs a recognizable valid access. An access can be
unidentified only to the extent that it is rooted in some non-address
type and we can prove that it is at least *not* part of an access to a
nominal class or global property. Pointer provenance is also required
for future IRGen-level bitfield optimizations.

Access analysis must be complete over address users: for an identified
object root all memory accesses including subobjects must be
discoverable.

Access analysis must be symmetric: use-def and def-use analysis must
be consistent.

AccessPath is merely a wrapper around the existing accessed-storage
utilities and IndexTrieNode. Existing passes already very succesfully
use this approach, but in an ad-hoc way. With a general utility we
can:

- update passes to use this approach to identify memory access,
  reducing the space and time complexity of those algorithms.

- implement an inexpensive on-the-fly, debug mode address lifetime analysis

- implement a lightweight debug mode alias analysis

- ultimately improve the power, efficiency, and maintainability of
  full alias analysis

- make our type-based alias analysis sensistive to the access path

atrick committed:
alias-analysis test

atrick committed:
RLE tests for boundary Projection indices.

atrick committed:
Update LoadCopyToLoadBorrow for MemAccessUtils API.

atrick committed:
AccessedStorage and AccessPath documentation.

atrick committed:
Cache 'isLet' within AccessedStorage.

Compute 'isLet' from the VarDecl that is available when constructing
AccessedStorage so we don't need to recover the VarDecl for the base
later.

This generally makes more sense and is more efficient, but it will be
necessary when we look past class casts when finding the reference root.

atrick committed:
Add AccessedStorageWithBase to conviently recover the base's VarDecl

atrick commented:
Discussion in https://github.com/apple/swift/pull/33121

atrick commented:
@swift-ci test

atrick commented:
@swift-ci test source compatibility

eeckstein commented on docs/SILProgrammersManual.md:
```diff
@@ -159,6 +159,401 @@ rules out PartialApplies is no excuse to conflate applied arguments
+AccessedStorage objects are identical.
+
+`AccessedStorage` records the "root" SILValue of the access. The root is
+the same as the access base for all storage kinds except global and
```
This is actually a very important sentence ("The root is the same as the base, except..."). Maybe it's worth noting that already in the first section.

atrick replied:
Added earlier
```
#### Access base

The access base is the address that directly identifies the kind of
storage being accessed, such as `alloc_box`, `alloc_stack`,
`global_addr`, `ref_element_addr`, and function arguments (see
`AccessedStorage::Kind`).

The access base is the same as the "root" SILValue for all storage
kinds except global and class storage. Global storage has no root. For
class storage the root is the SILValue that identifies object,
described as the "reference root" above.
```

eeckstein commented on docs/SILProgrammersManual.md:
```diff
@@ -159,6 +159,401 @@ rules out PartialApplies is no excuse to conflate applied arguments
+to reliably identify all uses of that storage location within the
+function for a particular access base. If the storage is uniquely
+identified, then that also implies that all uses of that storage
+within the function have been discovered.
```
To clarify, you could add "In other words, there are no aliases to the use set within the function"

atrick replied:
Yes, I also added this comment
```
As passes begin to adopt AccessPath::collectUses(), I expect it to
become a visitor pattern that allows the pass to perform custom
book-keeping for certain types of uses.
```

eeckstein reviewed (commented):
The documentation is much better now, thanks!
I still have a few comments.
And I still have to review the implementation.

atrick commented on docs/SILProgrammersManual.md:
```diff
@@ -159,6 +159,401 @@ rules out PartialApplies is no excuse to conflate applied arguments
+## `AccessedStorage` and `AccessPath`
+
+The `AccessedStorage` and `AccessPath` types formalize memory access
+in SIL. Given an address-typed SIL value, it must be possible to
```
done

atrick commented on docs/SILProgrammersManual.md:
```diff
@@ -159,6 +159,401 @@ rules out PartialApplies is no excuse to conflate applied arguments
+
+1. verify exclusivity enforcement
+
+2. continue optimizing exclusivity checks following other transforms
```
done

atrick commented on docs/SILProgrammersManual.md:
```diff
@@ -159,6 +159,401 @@ rules out PartialApplies is no excuse to conflate applied arguments
+
+Each object property and its tail storage is considered a separate
+formal access base. The reference root is only one component of an
+`AccessedStorage` location, its property index is another.
```
```
Each object property and its tail storage is considered a separate
formal access base. The reference root is only one component of an
`AccessedStorage` location. AccessedStorage also identifies the class
property being accessed within that object.
```

atrick commented on docs/SILProgrammersManual.md:
```diff
@@ -159,6 +159,401 @@ rules out PartialApplies is no excuse to conflate applied arguments
+accessed variable. The sequence of access projections between the base
+and the memory address correspond to an access path.
+
+For example, struct fields are not directly accessed, unless a nested
```
done

atrick commented on docs/SILProgrammersManual.md:
```diff
@@ -159,6 +159,401 @@ rules out PartialApplies is no excuse to conflate applied arguments
+completely.
+
+Access def-use chains are divided by three points: the "root", the
+formal access "base", and the address of a memory operation. For
```
```
Access def-use chains are divided by four points: the "root", the
access "base", the outer-most "access" scope, and the "address" of a
memory operation. For example:
```
And later on...
```
#### Access projections

On the def-use chain between the outermost formal access scope within
the current function and a memory operation, *access projections*
identify subobjects laid out within the formally accessed
variable.
```

atrick commented on docs/SILProgrammersManual.md:
```diff
@@ -159,6 +159,401 @@ rules out PartialApplies is no excuse to conflate applied arguments
+
+The first part of the def-use chain computes the formal access base
+from the root of the object (e.g. `alloc_ref ->
+ref_element_addr`). Only reference types or Builtin.BridgeObject types
```
So you know, the implementation of reference roots is not in this PR because I want to enable verification one step at a time without breaking too much.

```
The first part of the def-use chain computes the formal access base
from the root of the object (e.g. `alloc_ref ->
ref_element_addr`). The reference root might be a locally allocated
object, a function argument, a function result, or a reference loaded
from storage. There is no enforcement on the type of operation that
can produce a reference; however, only reference types or
Builtin.BridgeObject types are only allowed in this part of the
def-use chain. The reference root is the greatest common ancestor in
the def-use graph that can identify an object by a single SILValue. If
the root as an `alloc_ref`, then it is *uniquely identified*. The
def-use chain from the root to the base may contain reference casts
(`isRCIdentityPreservingCast`) and phis.
```

atrick commented on docs/SILProgrammersManual.md:
```diff
@@ -159,6 +159,401 @@ rules out PartialApplies is no excuse to conflate applied arguments
+certain conditions.
+
+For example, the following is a valid def-use access chain, with an
+access base defined in `b0`, a memory operation in `b3` and an
```
done

atrick commented on docs/SILProgrammersManual.md:
```diff
@@ -159,6 +159,401 @@ rules out PartialApplies is no excuse to conflate applied arguments
+#### Access base
+
+The access base is the address that directly identifies the kind of
+storage being accessed, such as `alloc_box`, `alloc_stack`,
```
Good question. I added this paragraph
```
"Box" storage is uniquely identified by an `alloc_box`
instruction. Therefore, we consider the `alloc_box` to be the base of
the access. Box storage does not apply to all box types or box
projections, which may instead originate from arguments or indirect
enums for example.
```

atrick replied:
Also, earlier:
```
The access base is the SILValue produced by an instruction that
directly identifies the kind of storage being accessed without further
use-def traversal. Common access bases are `alloc_box`, `alloc_stack`,
`global_addr`, `ref_element_addr`, and function arguments (see
`AccessedStorage::Kind`).
```

atrick commented on docs/SILProgrammersManual.md:
```diff
@@ -159,6 +159,401 @@ rules out PartialApplies is no excuse to conflate applied arguments
+#### Nested access
+
+Nested access occurs when an access base is a function argument. The
+caller always checks `@inout` arguments for exclusivity. However, the
```
Yes
```
Nested access occurs when an access base is a function argument. The
caller always checks `@inout` arguments for exclusivity (an access
marker must exist in the caller). However, the argument itself is a
variable with its own formal access.
```

atrick commented on docs/SILProgrammersManual.md:
```diff
@@ -159,6 +159,401 @@ rules out PartialApplies is no excuse to conflate applied arguments
+sil @takesTwoInouts : $@convention(thin) (@inout Int, @inout Int) -> ()
+
+sil @hasNestedAccess : $@convention(thin) (@inout Int) -> () {
+bb0(%0 : $*Int):
```
Conflicts may occur in the callee which were not evident in the caller. In this example, a conflict
occurs in `hasNestedAccess` but not in its caller:

```
func takesTwoInouts(_ : inout Int, _ : inout Int) -> () {}

func hasNestedAccess(_ x : inout Int) -> () {
  takesTwoInouts(&x, &x)
}

var x = 0
hasNestedAccess(&x)
```

Produces these access markers:
```
sil @takesTwoInouts : $@convention(thin) (@inout Int, @inout Int) -> ()

sil @hasNestedAccess : $@convention(thin) (@inout Int) -> () {
bb0(%0 : $*Int):
  %innerAccess  = begin_access [modify] %0 : $*Int
  %conflicting  = begin_access [modify] %0 : $*Int
  %f = function_ref @takesTwoInouts
  apply %f(%innerAccess, %conflicting)
    : $@convention(thin) (@inout Int, @inout Int) -> ()
  end_access %conflicting : $*Int
  end_access %innerAccess : $*Int
  //...
}

%var = alloc_stack $Int
%outerAccess  = begin_access [modify] %var : $*Int
%f = function_ref @hasNestedAccess
apply %f(%outerAccess) : $@convention(thin) (@inout Int) -> () {
end_access %outerAccess : $*Int
```

atrick commented on docs/SILProgrammersManual.md:
````diff
@@ -159,6 +159,401 @@ rules out PartialApplies is no excuse to conflate applied arguments
+end_access %outerAccess : $*Int
+```
+
+Nested accesses become part if the def-use chain after inlining.
````
Relevant updated sections:
```
Nested accesses become part if the def-use chain after inlining. Here,
both `%innerAccess` and `%conflicting` are nested within
`%outerAccess`:
```

```
%var = alloc_stack $Int
%outerAccess  = begin_access [modify] %var : $*Int
%innerAccess  = begin_access [modify] %outerAccess : $*Int
%conflicting  = begin_access [modify] %outerAccess : $*Int
%f = function_ref @takesTwoInouts
apply %f(%innerAccess, %conflicting)
  : $@convention(thin) (@inout Int, @inout Int) -> ()
end_access %conflicting : $*Int
end_access %innerAccess : $*Int
end_access %outerAccess : $*Int
```

```
For most purposes, the inner access scopes are irrelevant. When we ask
for the "accessed storage" for `%innerAccess`, we get an
`AccessedStorage` value of "Stack" kind with base `%var =
alloc_stack`. If instead of finding the original accessed storage, we
want to identify the enclosing formal access scope, we need to use a
different API that supports the special `Nested` storage kind. This is
typically only used for exclusivity diagnostics though.

TODO: Nested static accesses that result from inlining could
potentially be removed, as long as DiagnoseStaticExclusivity has
already run.

#### Access projections

On the def-use chain between the *outermost* formal access scope within
the current function and a memory operation, *access projections*
identify subobjects laid out within the formally accessed
variable. The sequence of access projections between the base and the
memory address correspond to an access path.

For example, there is no formal access for struct fields. Instead,
they are addressed using a `struct_element_addr` within the access
scope:
```

```
%access  = begin_access [read] [static] %base : $*S
%memaddr = struct_element_addr %access : $*S, #.field
%value   = load [trivial] %memaddr : $*Int64
end_access %access : $*S
```

```
Note that is is possible to have a nested access scope on the address
of a struct field, which may show up as an access of
struct_element_addr after inlining.
```

atrick commented:
@swift-ci test

atrick commented:
@swift-ci test

eeckstein commented on lib/SIL/Utils/MemAccessUtils.cpp:
```diff
@@ -228,200 +595,748 @@ void AccessedStorage::print(raw_ostream &os) const {
+    if (storage.getKind() == AccessedStorage::Class) {
+      pathIndices.push_back(AccessPath::Index::forSubObjectProjection(
+          storage.getPropertyIndex()));
+    }
```
} else if...

atrick committed:
MemAccessUtils comment

atrick committed:
Temporarily disable the load-borrow checker.

A rewrite is ready and will be merged ASAP.

atrick committed:
Add AccessUseVisitor and cleanup related APIs.

Add AccesssedStorage::compute and computeInScope to mirror AccessPath.

Allow recovering the begin_access for Nested storage.

Adds AccessedStorage.visitRoots().

atrick commented:
@gottesmm In order to stage in new verification, load-borrow immutability gets disabled in this PR but immediately reenabled here: https://github.com/atrick/swift/commits/rewrite-loadborrowchecker

atrick commented:
@swift-ci test

atrick commented:
@swift-ci test source compatibility

atrick commented:
@swift-ci benchmark

atrick commented:
@swift-ci test

atrick commented:
@swift-ci test source compatibility

gottesmm commented:
@atrick that's a huge patch. If you are going to disable it, now that if anything creeps in, you need to fix.

atrick commented:
@gottesmm I've actually been waiting to land this PR until I know all the subsequent functionality, including the load-borrow-checker rewrite will pass all testing and SCK. So the only reason to disable load borrow checking here is to break up the functionality and allow the bots to churn through one piece at a time.

gottesmm commented:
Sure. I am cool with it. I am just saying if something sneaks in... its on you to fix = p.

atrick committed:
The verifier should only output to llvm::errs()

atrick commented:
@swift-ci smoke test
